spec/core/ics-004-channel-and-packet-semantics/README.md

```diff
@@ -626,14 +745,23 @@ function recvPacket(
     abortTransactionUnless(connection !== null)
     abortTransactionUnless(connection.state === OPEN)
 
-    abortTransactionUnless(connection.verifyPacketData(
-      proofHeight,
-      proof,
-      packet.sourcePort,
-      packet.sourceChannel,
-      packet.sequence,
-      hash(packet.data, packet.timeoutHeight, packet.timeoutTimestamp)
-    ))
+    if (len(channel.connectionHops) > 1) {
+      consensusState = provableStore.get(consensusStatePath(connection.ClientId, proofHeight))
+      key = host.PacketCommitmentPath(packet.GetDestPort(), packet.GetDestChannel(), packet.GetSequence())
+      commitment = types.CommitPacket(packet)
+      counterpartyConnectionEnd = abortTransactionUnless(getMultihopConnectionEnd(proof))
+      prefix = counterpartyConnectionEnd.GetCounterparty().GetPrefix()
+      abortTransactionUnless(VerifyMultihopProof(k.cdc, consensusState, channel.ConnectionHops, proof, prefix, key, commitment))
+    } else {
+      abortTransactionUnless(connection.verifyPacketData(
```

Comment:
Was looking at what we're missing for multi-hop by not calling this (not using the client security model):
- clients on intermediate chains are not active
- not checking delay for packet delay feature
Same checks are skipped in other packet handlers.

Reply:
Same for channel handlers.

Reply:
Yes the active client is a great point. If an intermediate client gets frozen for misbehaviour, we won't detect it here. This could cause a proof to succeed against the forked consensus state. We should definitely be checking that all intermediate clients are active.

I mentioned the packet delay feature to the team. I think it makes sense to only allow `packet_delay_time` to be specified in this case, and for the intermediate consensus states to have their timestamps checked against the local clock of the final destination chain. Rather than checking against all the intermediate BFT clocks

Reply:
> I mentioned the packet delay feature to the team. I think it makes sense to only allow packet_delay_time to be specified in this case, and for the intermediate consensus states to have their timestamps checked against the local clock of the final destination chain. Rather than checking against all the intermediate BFT clocks

Ok just to clarify, with multihop we only take into account the last connection delay and the pseudocode above will change to perform some version of `verifyDelayPeriodPassed`, in addition to the checks you mention. 
So a relayer would send all client updates, collect the proofs and, if the last connection delay is non-zero, wait for the delay of the last hop before sending the packet.

Regarding the checks you mention, we need to allow for clock drift and block time variations. There is also an accumulated delay for the client update on the destination chain since for every hop the relayer needs to wait for the client update to be sent and consensus state to be stored on chain before we do the next chain update. So the timestamp of consensus state on chain `i` may be `blockTime(i)` older than the consensus state on chain `i+1`. This may be event higher if the update is not included in the "next" block.

Note: even with zero delay packets will be effectively relayed with a delay of `sum(blockTime(i), i = 1..N-1)`

Not sure if all this is accurate, just some thoughts and you may have a different view or I may miss something obvious. To clarify any confusion, could we add more english explanations to the specs to make things clear.

Reply:
> Ok just to clarify, with multihop we only take into account the last connection delay and the pseudocode above will change to perform some version of verifyDelayPeriodPassed, in addition to the checks you mention.

Or do we delay the packet/ client updates with the sum of the delays of all previous connections? Each hop delay is available to the destination chain so it could verify that consensus state timestamps of adjacent chains respect the connection delay.

Reply:
Agreed that more English explanations in addition to the pseudocode would be very useful.

> Ok just to clarify, with multihop we only take into account the last connection delay and the pseudocode above will change to perform some version of verifyDelayPeriodPassed, in addition to the checks you mention.
So a relayer would send all client updates, collect the proofs and, if the last connection delay is non-zero, wait for the delay of the last hop before sending the packet.

It would wait such that that the client updates of all intermediate hops have at least passed the delay period specified at the final hop. Yes, I think in practice this would mean waiting for last client update to pass the delay period most of the time. But the point is the delay period parameter specified on the conneciton end will be enforced against every intermediate consensus state using the BFT clock of the final proving chain.

> Regarding the checks you mention, we need to allow for clock drift and block time variations. There is also an accumulated delay for the client update on the destination chain since for every hop the relayer needs to wait for the client update to be sent and consensus state to be stored on chain before we do the next chain update. So the timestamp of consensus state on chain i may be blockTime(i) older than the consensus state on chain i+1. This may be even higher if the update is not included in the "next" block.

I don't believe this is as relevant. Perhaps in order to route a packet, a relayer may have to submit client updates on some intermediate hops. But if all the hops are up-to-date, a relayer can simply query proofs from each chain and submit them. So the minimum delay is negligible, and the maximum delay as you said is `sum(blockTime(i))`

The purpose of the block delay period is to ensure that there is some grace period for detecting misbehaviour before we use the consensus state for proof verification.

The clock drift is not cumulative across the path, instead the effective clock drift on this connection will be the maximum drift between any two chains on this route.

So supposing we wanted to allow for a grace period of time `n` for relayers to submit misbehaviour against a consensus state. We would want the delay period to be something like:

```
delayPeriodTime = n + sumBlock(i) + max_clock_drift
```

In practice, I think in order for `n` to be a useful grace period at all, `n >> sumBlock(i) + maxClockDrift`.

So rather than explicitly accounting for this in the protocol, I think if a user wants delayPeriod protection they should just set the value to be large enough to account for any of the above factors. This should be explained in the spec

Reply:
Ahh wait, I just realized the delayPeriod is implemented to be enforced against time of addition not the actual timestamp of the consensus state. So it is not easy to do this in the multihop case...

Will have to think about it some more

Reply:
> This can happen if the later hops clients were more recently updated by other relayers and only the first couple hops needed to have updates filled in to complete the route.

This statement and other comments in the spec suggest that I am missing something very basic. How can this be? If you update a client at chain `i` before you update the client at chain `i-1` the recursive proof at chain `i` will fail, no? Since the commitment root stored is for an app state on `i-1` that did not include the the expected commitment (consensus state for `i-2`).

I thought that the consensus state on chain `i` must be created after the one in `i-1`, otherwise the proof chain is broken. Is this not true? Also, the commitment proof on chain `i-1` must be taken at the exact height as the height of the consensus state (that provides the commitment root) on chain `i`.

Reply:
> Ahh wait, I just realized the delayPeriod is implemented to be enforced against time of addition not the actual timestamp of the consensus state. So it is not easy to do this in the multihop case...
> 
> Will have to think about it some more
> 
> Especially since we do not include time of addition inside the consensus state itself

Ok but if:
- the destination chain `N` verifies the delay against the time it stored the consensus state for  `N-1`, and 
- it is true that `consensusState[i]` must have been stored after `consensusState[i-1]` for proofs to verify (see my last questions in https://github.com/cosmos/ibc/pull/882#discussion_r1107409522)

then the delayPeriod enforcement is also proven